docs/usage/operations.rst

```diff
@@ -1009,3 +1009,183 @@ Class Definitions
                   ),
               ),
           ]
+
+
+.. py:class:: SaferRenameModelPart1(old_name: str, new_name: str)
+
+    First step on a routine that provides a safer RenameModel alternative.
+
+    :param old_name: Old model name as it was once defined e.g Foo.
+    :type old_name: str
+    :param new_name: New model name as it is now defined e.g NewFoo.
+    :type new_name: str
+
+    **Why use this SaferRenameModelPart1 operation?**
+    -------------------------------------------------
+
+    When using Django's default ``RenameModel`` operation, the SQL created has
+    the following form:
+
+    .. code-block:: sql
+
+      BEGIN;
+      --
+      -- Rename model Foo to NewFoo
+      --
+      ALTER TABLE "myapp_foo" RENAME TO "myapp_newfoo";
+      COMMIT;
+
+    In modern applications that use gradual deployments like blue/green
+    deploys, renaming a table in-flight might cause issues when the traffic
+    hasn't fully been moved from the old servers (blue) to the new ones
+    (green).
+
+    For example, in a Django app the old servers would see the app state as if
+    the recently renamed table was still using the old name, and will therefore
+    crash when the model is used on those servers.
+
+    Additionally if the model has foreign keys, the relating models will have
+    their foreign key constraints created from scratch which might take a long
+    time. Column names for M2M tables are also updated. For example:
+
+    .. code-block:: py
+
+      class NewFoo(models.Model):
+          pass
+
+
+      class Bar(models.Model):
+          foo = models.ForeignKey(NewFoo, on_delete=models.CASCADE, null=True)
+
+
+      class Buzz(models.Model):
+          foos = models.ManyToManyField(NewFoo)
+
+    Culminates on the following statements:
+
+    .. code-block:: sql
+
+      BEGIN;
+      --
+      -- Rename model Foo to NewFoo
+      --
+      ALTER TABLE "myapp_foo" RENAME TO "myapp_newfoo";
+
+      SET CONSTRAINTS "myapp_bar_foo_id_f5927bae_fk_myapp_newfoo_id" IMMEDIATE;
+
+      ALTER TABLE "myapp_bar" DROP CONSTRAINT "myapp_bar_foo_id_f5927bae_fk_myapp_newfoo_id";
+
```

Comment:
Why would the constraint name `myapp_bar_foo_id_f5927bae_fk_myapp_newfoo_id` include "newfoo" at this point? (I suspect that this example has been put together by hand.)

From [here](https://github.com/kraken-tech/django-pg-migration-tools/pull/74#issuecomment-2632216141):

> I think that the problem here is that Django actually does more than it should.
Namely, dropping constraints and recreating them from scratch so that they can
have a fresh name, that is costly for FKs. In this case, that's explained in
the commit that adds the docs here which includes the output of the DDLs
performed by Django. Hope that clarifies?

I would expect to see a different constraint name after Django drops the old one and makes a new one. So no, the example here doesn't currently clarify :(